File: VB/NorthwindDataProvider.vb
Imports System.Data
Imports System.Data.SqlClient

Namespace RichEditMasterDetailMailMerge

    Public Module NorthwindDataProvider

        Private connectionString As String = "Data Source=.\SQLExpress2008;Initial Catalog=Northwind;Integrated Security=SSPI"

        Sub New()
            Using connection As SqlConnection = New SqlConnection(connectionString)
                Dim selectCommand As SqlCommand = New SqlCommand("SELECT * FROM Categories", connection)
                Dim da As SqlDataAdapter = New SqlDataAdapter(selectCommand)
                categoriesField = New DataTable("Categories")
                productsField = New DataTable("Products")
                da.Fill(categoriesField)
                selectCommand.CommandText = "SELECT * FROM Products"
                da.Fill(productsField)
                categoriesField.Constraints.Add("IDPK", categoriesField.Columns("CategoryID"), True)
                productsField.Constraints.Add("IDPK", productsField.Columns("ProductID"), True)
                selectCommand.Dispose()
            End Using
        End Sub

        Private categoriesField As DataTable

        Public ReadOnly Property Categories As DataTable
            Get
                Return categoriesField
            End Get
        End Property

        Private productsField As DataTable

        Public ReadOnly Property Products As DataTable
            Get
                Return productsField
            End Get
        End Property

        Public Function GetProductsByCategoryId(ByVal categoryId As Integer) As DataTable
            Dim dataView As DataView = New DataView(Products, String.Format("CategoryID = {0}", categoryId), String.Empty, DataViewRowState.CurrentRows)
            Return dataView.ToTable("ProductsOfCategory")
        End Function
    End Module
End Namespace

File: CS/NorthwindDataProvider.cs
using System.Data;
using System.Data.SqlClient;

namespace RichEditMasterDetailMailMerge {
    public static class NorthwindDataProvider {
        private static string connectionString = @"Data Source=.\SQLExpress;Initial Catalog=Northwind;Integrated Security=SSPI";

        static NorthwindDataProvider() {
            using (SqlConnection connection = new SqlConnection(connectionString)) {
                SqlCommand selectCommand = new SqlCommand("SELECT * FROM Categories", connection);
                SqlDataAdapter da = new SqlDataAdapter(selectCommand);

                categories = new DataTable("Categories");
                products = new DataTable("Products");

                da.Fill(categories);
                selectCommand.CommandText = "SELECT * FROM Products";
                da.Fill(products);

                categories.Constraints.Add("IDPK", categories.Columns["CategoryID"], true);
                products.Constraints.Add("IDPK", products.Columns["ProductID"], true);
 
                selectCommand.Dispose();
            }
        }

        private static DataTable categories;

        public static DataTable Categories {
            get {
                return categories;
            }
        }

        private static DataTable products;

        public static DataTable Products {
            get {
                return products;
            }
        }

        public static DataTable GetProductsByCategoryId(int categoryId) {
            DataView dataView = new DataView(Products, string.Format("CategoryID = {0}", categoryId), string.Empty, DataViewRowState.CurrentRows);

            return dataView.ToTable("ProductsOfCategory");
        }
    }
}